Unit tests and modularisation of YieldGeneratorBasic

domob1812 opened the pull request:
This extends the yield-generator code.  It adds basic unit tests for `YieldGeneratorBasic` (there were none so far), and also makes the logic a bit more "modular" and extensible.  The functions that determine how funds are sent between mixdepths are moved out of `create_my_orders` and `oid_to_order`, so that custom YG scripts can more easily change them without having to duplicate all the more general logic in those functions.

With this, it will be trivial to implement a custom YG logic as described in #404, but also other custom scripts in the future.  At the moment, those (e.g. https://github.com/JoinMarket-Org/custom-scripts/blob/master/yield-generators/yg-randomizer.py) just copy the full code for `create_my_orders` and `oid_to_order` and tweak it, which is bad for code readability and maintainability.

domob1812 commented:
See https://github.com/JoinMarket-Org/custom-scripts/pull/4 for how this can be used to more easily and with less code implement custom YG scripts.

domob1812 commented:
Hm, that Travis failure is weird.  When I run `jmbase/test/test_commands.py` locally, it succeeds for me.  Also I currently don't see any connection at all of that test to my changes.  Any hints what it could be?

Also, even if that test is failing right now, I'm happy to get any feedback/review of the PR.

AdamISZ commented on jmclient/test/test_yieldgenerator.py:
```diff
@@ -0,0 +1,196 @@
+from jmclient import jm_single, YieldGeneratorBasic
+
+
+class FakeWallet:
```
Just to mention, we already have DummyWallet as a test class.

undeath replied:
Yes, I'd prefer using the existing `DummyWallet`, too. Implementing dedicated mock wallets for each test will just cause considerable overhead when changing something.

domob1812 replied:
I know, I've seen the `DummyWallet`.  However, it seemed to have "just" a bunch of hardcoded UTXOs in it, and I didn't see an easy way to a) set the balances of each mixdepth from the test, and b) e.g. verify that the UTXOs or addresses returned are from a certain mixdepth.  Maybe I just missed how to do it.

In other words, it seemed to me that `DummyWallet` was much more geared towards testing lower-level taker/maker stuff, rather than the abstract level of the yield generator.

undeath replied:
Indeed, with the `DummyWallet` from `test_taker` it's actually hard to tell how it's going to behave outside of that test suite. Is anything wrong with using the actual wallet implementation?

domob1812 replied:
I've taken a brief look at the actual wallet, and it seems like we could make it do all things required for the test relatively easily.  I'll try that out later and will switch to it if it seems possible.

domob1812 replied:
The test is now updated to use a real wallet (with some tiny utility functions added on top).

AdamISZ commented:
Another somewhat peripheral comment (I haven't read the code yet): we use pytest and it can handle this style of test, but I'd prefer to stick with the pattern we use in all other tests, just test_* functions with fixtures and parametrization where useful.

undeath commented:
Thanks for adding yg tests, those are really useful!

The test fail is related to an unclean reactor: https://travis-ci.org/JoinMarket-Org/joinmarket-clientserver/jobs/590736375#L2113

If some code run in your test starts/uses reactor you have to tear it down like done here: https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/test/test_coinjoin.py#L314-L317

undeath commented on jmclient/jmclient/yieldgenerator.py:
```diff
@@ -159,6 +161,28 @@ def on_tx_confirmed(self, offer, confirmations, txid):
                 confirm_time / 60.0, 2), ''])
         return self.on_tx_unconfirmed(offer, txid, None)
 
+    def get_available_mixdepths(self, verbose=True):
```
I think it should be `verbose=False` by default. `verbose` is not implemented right now but it would probably add some logging output that would usually not be needed, especially for a yg.

domob1812 replied:
I'm happy to change it, but I just used what the wallet function itself has: https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/3af6f43da280fed3077f715848cc3b6d2f1f32d1/jmclient/jmclient/wallet.py#L703

undeath replied:
Ah, yes, I see. I still think for YG it should be `False` by default.

domob1812 replied:
Ok.  I've set it to false now everywhere in the YG.

domob1812 commented:
> Another somewhat peripheral comment (I haven't read the code yet): we use pytest and it can handle this style of test, but I'd prefer to stick with the pattern we use in all other tests, just test_* functions with fixtures and parametrization where useful.

Yes, I've seen that.  The testcase classes are not really necessary, but I think they help to group the tests in a logical way (all tests for `create_my_orders`, all tests for `oid_to_order` and so on).  But if you prefer not to do that, I can just have a bunch of flat functions e.g. with prefixes in the names to express that instead.

> If some code run in your test starts/uses reactor you have to tear it down like done here: https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/test/test_coinjoin.py#L314-L317

I've seen the error about reactors.  But to be honest, I don't know what a reactor is, nor do I know if my tests use one (so presumably not, unless it is hidden somewhere).

domob1812 commented:
I've done some more testing, and it seems that "unclean reactor" failure is somehow due to an interaction between the test environments of `test_yieldgenerator.py` and `test_commands.py`.  When I just run three tests, `test_wallet.py`, `test_yieldgenerator.py` and `test_commands.py`, the error occurs.  If I change the order to `test_yieldgenerator.py`, `test_wallet.py` and `test_commands.py`, then all succeeds.  (And all of that is true even with just my first commit, which only adds the new test and does not do any other changes at all.)

So either `test_yieldgenerator.py` is somehow messing up the environment, or `test_commands.py` is somehow relying on the environment from `test_wallet.py` (which it presumably shouldn't).

Does that ring a bell, what might be going wrong?  I'll try to find more information by myself as well.

domob1812 commented:
Seems my previous comment was a bit too soon - I've found the issue now with more analysis, and fixed it.  Tests are passing now.

@AdamISZ @undeath:  I've replied to your review comments why I wrote the code the way it is, which I still think makes sense.  But since this is of course your decision in the end, please take a look and let me know if you want me to change things (in view of my rationale for them).  Thanks!

undeath commented:
Please don't force-push updates until the code has been fully reviewed. It makes it very hard to know what has been changed since the last review, forcing every reviewer to start from scratch.

domob1812 commented:
> Please don't force-push updates until the code has been fully reviewed. It makes it very hard to know what has been changed since the last review, forcing every reviewer to start from scratch.

Ok.  I've done refinements now as additional small commits.  I will rebase those to get a clean set of commits when the review is complete.

undeath commented on jmclient/jmclient/yieldgenerator.py:
```diff
@@ -93,23 +91,24 @@ def create_my_orders(self):
-        filtered_mix_balance = [m
-                                for m in iteritems(mix_balance)
-                                if m[1] >= total_amount]
+        filtered_mix_balance = {m: b
```
`filtered_mix_balance` could be moved to `select_input_mixdepth`. Seems cleaner to me.

domob1812 replied:
The reason it is not there is that filtering for available mixdepths is a generic task, while selecting the one to use from them is something that may depend on the custom algorithm.  And splitting the generic things from the actual selection logic is the main goal of this change, thus I think we should keep as much generally useful logic in `oid_to_order` as possible.  (Or put another way, `select_input_mixdepth` should need to do as little as possible, as that is what custom algorithms will override and reimplement.)

undeath commented on jmclient/test/test_yieldgenerator.py:
```diff
@@ -0,0 +1,205 @@
+        """Asserts that the list of UTXOs (as returned from UTXO selection
+        of the wallet) is all from the given expected mixdepth."""
+
+        for _, u in iteritems(utxos):
```
minor nit, this should be `for u in utxos.values():`

domob1812 replied:
Ah yes, Python is not my main language.  Thanks for pointing this out!

undeath commented:
Thanks for the change. Just a note, we have a function `make_wallets` in `commontest.py` but that would require the test to interact with the bitcoind and spin up the whole twisted network, which is not really needed here. So I think your solution to slightly customise the wallets behaviour is fine.

If you rebase+stash this can be merged.

domob1812 commented:
Ok cool, thanks for the review!  I've squashed the commits and rebased.  (Commit 3f02b0b88c00563a059cabac50d0e29b3e797e08 is just that.)

However, the upstream change to `on_tx_unconfirmed` somehow made the tests for it fail.  I'll debug this now.

undeath commented:
Changes look good to me. Can you rename the `TestWallet` class to something different? pytest emits a warning because it thinks that's a test suite since the class starts with "Test":

    clientserver/jmclient/test/test_yieldgenerator.py:12: PytestCollectionWarning: cannot collect test class 'TestWallet' because it has a __init__ constructor (from: jmclient/test/test_yieldgenerator.py)

domob1812 commented:
Ah yes, good point.  I've renamed it to `CustomUtxoWallet` and squashed the commits again.

undeath commented:
merged in 3966810

AdamISZ commented on scripts/yg-privacyenhanced.py:
```diff
@@ -35,12 +35,10 @@
 
     def create_my_orders(self):
-        mix_balance = self.wallet.get_balance_by_mixdepth()
+        mix_balance = self.get_available_mixdepths(verbose=False)
```
This appears to be a bug? `YieldGenerator.get_available_mixdepths()` has no keyword argument `verbose`.

AdamISZ replied:
Sorry I meant `YieldGeneratorBasic` not `YieldGenerator` but same comment.

undeath replied:
yes, indeed!

domob1812 replied:
Ah yes, I apparently missed that when I updated the PR for review comments.  Sorry - fix is in #410.
